sympy/sets/handlers/intersection.py

```diff
@@ -277,9 +277,9 @@ def intersection_sets(self, other): # noqa:F811
                 return FiniteSet(*(fn.subs(n, s[0]) for s in solns))
 
     if other == S.Reals:
-        from sympy.solvers.solveset import solveset_real
```

Comment:
Although this PR can be patched to use `solve_linear` isn't it better to use the full `solveset_real`? Or maybe you are thinking to keep that for the `simplify` method? But as of now, `simplify` doesn't do anything to the ConditionSet.

In addition, you are incorrectly passing args to `solve_linear`; check parameters.

Reply:
Based on the limited knowledge I have, I thought using solveset_real would be computationally expensive. Although using solveset_real would make this implementation a lot simpler

Reply:
@smichr Should I use solveset_real or is this cool? I've fixed the other stuff

Reply:
There is no choice but to use `solveset_real` or else even a simple system's intersection won't be computed. Is there another method that will resolve this simple failing case? `solve_linear` is ok at instantiation where we don't want to work too hard, but for the `intersection` method, we have to do our best to give a solution.

And you are still not using `solve_linear` correctly. Try solving a simple equation like `x + 1 = 0` after reading the docstring and confirm that you get `(x, -1)` as the solution.

Reply:
A function like `solveset_real` can be very slow in some cases. I don't think that intersection or instantiation should call a potentially slow function like this. It is actually especially bad in something like an intersection handler because these handlers are called for all *pairs* of sets in an intersection. That makes computing an intersection of a large number of sets very slow e.g.:
```python
In [9]: s = [ImageSet(Lambda(n, x**n), Integers) for x in range(2, 100)]

In [10]: Intersection(*s)
^C---------------------------------------------------------------------------
KeyboardInterrupt  
```
If there is a need to call `solveset` then the whole operation should be handled as part of simplification rather than in any part of evaluation. The intersect handlers need to be extremely fast because they are themselves called in an algorithmically bad `O(n**2)` way.

Reply:
OK. Then we need to add this to `simplify` at some point. As for the issue here, it is easy to make `solve_linear` still work for the given example by arg-wise calling `solve_linear`:
```diff
diff --git a/sympy/sets/handlers/intersection.py b/sympy/sets/handlers/intersection.py
index 2a34869063..3691b864eb 100644
--- a/sympy/sets/handlers/intersection.py
+++ b/sympy/sets/handlers/intersection.py
@@ -1,5 +1,5 @@
 from sympy import (S, Dummy, Lambda, symbols, Interval, Intersection, Set,
-                   EmptySet, FiniteSet, Union, ComplexRegion)
+                   EmptySet, FiniteSet, Union, ComplexRegion, Mul)
 from sympy.multipledispatch import dispatch
 from sympy.sets.conditionset import ConditionSet
 from sympy.sets.fancysets import (Integers, Naturals, Reals, Range,
@@ -304,16 +304,16 @@ def intersection_sets(self, other): # noqa:F811
             return None
         else:
             # univarite imaginary part in same variable
-            x, xis = solve_linear(im, n)
-            if x == n:
+            x, xis = zip(*[solve_linear(i, 0) for i in Mul.make_args(im) if n in i.free_symbols])
+            if x and all(i == n for i in x):
                 base_set -= FiniteSet(xis)
             else:
                 base_set -= ConditionSet(n, Eq(im, 0), S.Integers)
         # exclude values that make denominators 0
         for i in denoms(f):
             if i.has(n):
-                x, xis = solve_linear(i, n)
-                if x == n:
+                x, xis = zip(*[solve_linear(i, 0) for i in Mul.make_args(im) if n in i.free_symbols])
+                if x and all(i == n for i in x):
                     base_set -= FiniteSet(xis)
                 else:
                     base_set -= ConditionSet(n, Eq(i, 0), S.Integers)
```